### PR TITLE
Window UDF signature check

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #.

## Rationale for this change

We haven't add signature check for window udf, so we didn't get the expected error if the arguments failed to match the signature

### CODE REVIEW

File: datafusion/sqllogictest/test_files/window.slt [hunk 1]
```diff
@@ -4879,3 +4879,21 @@ SELECT lead(column2, 1.1) OVER (order by column1) FROM t;
...
+statement ok
+create table t(a int, b int) as values (1, 2)
+
+query II
+select a, row_number() over (order by b) as rn from t;
+----
+1 1
+
+# RowNumber expect 0 args.
+query error
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/window.slt:4895]: not an error on `main`

File: datafusion/expr/src/udwf.rs [hunk 2]
```diff
@@ -353,6 +358,29 @@ pub trait WindowUDFImpl: Debug + Send + Sync {
     fn sort_options(&self) -> Option<SortOptions> {
         None
     }
+
+    /// Coerce arguments of a function call to types that the function can evaluate.
+    ///
+    /// This function is only called if [`WindowUDFImpl::signature`] returns [`crate::TypeSignature::UserDefined`]. Most
+    /// UDAFs should return one of the other variants of `TypeSignature` which handle common
```
Comment (goldmedal) [inline datafusion/expr/src/udwf.rs:365]:
```suggestion
    /// UDWFs should return one of the other variants of `TypeSignature` which handle common
```
It looks like a typo (?

**Review by goldmedal — APPROVED** [review]
Thanks @jayzhan211 Overall LGTM but one typo comment.

File: datafusion/sqllogictest/test_files/window.slt (hunk 1 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/window.slt:4895]: Wild -- I double checked and you are right

```sql
DataFusion CLI v41.0.0
> create table t(a int, b int) as values (1, 2);
0 row(s) fetched.
Elapsed 0.015 seconds.

> select a, row_number() over (order by b) as rn from t;
+---+----+
| a | rn |
+---+----+
| 1 | 1  |
+---+----+
1 row(s) fetched.
Elapsed 0.006 seconds.

> select a, row_number(a) over (order by b) as rn from t;
+---+----+
| a | rn |
+---+----+
| 1 | 1  |
+---+----+
1 row(s) fetched.
Elapsed 0.002 seconds.
```

**Review by alamb — APPROVED** [review]
Thanks @jayzhan211  and @goldmedal for the review

I didn't realize there was no signature checking 🤯

**Comment by jayzhan211** [comment]
Thanks @goldmedal @alamb

**Pull request merged**